### PR TITLE
Add basic functionalities for StarkBank, StarkInfra and StarkSign SDKs

### PR DESCRIPTION
**Pull request by xavier-stark**

### CODE REVIEW

File: lib/utils/api.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+module StarkCore
+  module Utils
+    module API
+      def self.build_entity_hash(entity)
+        entity_hash = {}
+        entity_hash = entity if entity.is_a?(Hash)
+
+        entity.instance_variables.each do |key|
+          variable = entity.instance_variable_get(key)
+          entity_hash[key[1..-1]] = variable.is_a?(StarkCore::Utils::Resource) ? build_entity_hash(variable) : entity.instance_variable_get(key)
```
Comment (xavier-stark) [inline lib/utils/api.rb]: ver se cabe SubResource

File: lib/utils/checks.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+      def self.check_user(user)
+        return user if user.is_a?(StarkCore::User)
+        user = user.nil? ? StarkCore.user : user
+        raise(ArgumentError, 'A user is required to access our API. Check our README: https://github.com/starkbank/sdk-ruby/') if user.nil?
+        return user
+      end
+
+      def self.check_language
+      language = StarkCore.language
```
Comment (xavier-stark) [inline lib/utils/checks.rb]: indent

File: lib/utils/parse.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                            language: language, timeout: timeout, refresh: true)
+          return content
+        end
+
+        raise(StarkCore::Error::InvalidSignatureError, 'The provided signature and content do not match the Stark bank public key')
+      end
+
+      def self.verify_signature(content:, signature:, sdk_version:, host:, api_version:,
+                                user:, language:, timeout:, refresh: false)
+        public_key =  get_public_key_pem(
```
Comment (xavier-stark) [inline lib/utils/parse.rb]: 2 espaços

File: lib/utils/request.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        def json
+          JSON.parse(@content)
+        end
+      end
+
+      def self.fetch(host:, sdk_version:, user:, method:, path:, payload: nil, query: nil, 
+                      api_version: "v2", language: "en-US", timeout: 15)
+        user = Checks.check_user(user)
+        language = Checks.check_language
```
Comment (xavier-stark) [inline lib/utils/request.rb]: not checking language

File: lib/utils/request.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        url = "#{base_url}/#{path}#{StarkCore::Utils::URL.urlencode(query)}"
+        uri = URI(url)
+
+        access_time = Time.now.to_i
+        body = payload.nil? ? '' : payload.to_json
+        message = "#{user.access_id}:#{access_time}:#{body}"
+        signature = EllipticCurve::Ecdsa.sign(message, user.private_key).toBase64
+
+        case method
+        when 'GET'
```
Comment (xavier-stark) [inline lib/utils/request.rb]: indent

File: lib/utils/request.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+          req.body = body
+        else
+          raise(ArgumentError, 'unknown HTTP method ' + method)
+        end
+
+        req['Access-Id'] = user.access_id
+        req['Access-Time'] = access_time
+        req['Access-Signature'] = signature
+        req['Content-Type'] = 'application/json'
+        req['User-Agent'] = "Ruby-#{RUBY_VERSION}-SDK-#{host}-#{sdk_version}"
```
Comment (xavier-stark) [inline lib/utils/request.rb:71]: checar sdk python

File: lib/utils/rest.rb [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+          host: host,
+          sdk_version: sdk_version,
+          user: user,
+          method: 'POST',
+          path: StarkCore::Utils::API.endpoint(resource_name),
+          query: query,
+          payload: payload,
+          api_version: api_version,
+          timeout: timeout
+
```
Comment (xavier-stark) [inline lib/utils/rest.rb]: tirar linha

File: lib/utils/sub_resource.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module StarkCore
+    module Utils
```
Comment (xavier-stark) [inline lib/utils/sub_resource.rb]: 2 indent

File: lib/utils/api.rb [hunk 9]
```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+require_relative('case')
+
+module StarkCore
+  module Utils
+    module API
+      def self.build_entity_hash(entity)
+        entity_hash = {}
```
Comment (xavier-stark) [inline lib/utils/api.rb:9]: 2 indent

File: lib/utils/api.rb [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+      def self.api_json(entity)
+        built_hash = build_entity_hash(entity)
+        cast_json_to_api_format(built_hash)
+      end
+
+      def self.cast_json_to_api_format(hash)
+        entity_hash = {}
+        hash.each do |key, value|
+        next if value.nil?
```
Comment (xavier-stark) [inline lib/utils/api.rb]: indent

**Review by cdottori-stark — APPROVED** [review]

**Pull request merged**